crates/api_crud/src/user/create.rs

```diff
@@ -238,3 +227,319 @@ pub async fn register(
 
   Ok(Json(login_response))
 }
+
+#[tracing::instrument(skip(context))]
+pub async fn authenticate_with_oauth(
+  data: Json<AuthenticateWithOauth>,
+  req: HttpRequest,
+  context: Data<LemmyContext>,
+) -> LemmyResult<Json<LoginResponse>> {
+  let site_view = SiteView::read_local(&mut context.pool())
+    .await?
+    .ok_or(LemmyErrorType::LocalSiteNotSetup)?;
+
+  let local_site: LocalSite = site_view.local_site.clone();
+
+  // validate inputs
+  if data.oauth_provider_id == OAuthProviderId(0) || data.code.is_empty() || data.code.len() > 300 {
+    return Err(LemmyErrorType::OauthAuthorizationInvalid)?;
+  }
+
+  // validate the redirect_uri
+  let redirect_uri = &data.redirect_uri;
+  if redirect_uri.host_str().unwrap_or("").is_empty()
+    || !redirect_uri.path().eq(&String::from("/oauth/callback"))
+    || !redirect_uri.query().unwrap_or("").is_empty()
+  {
```

Comment:
as far as I know, having a whitelist of redirect_uris is not necessary for the oauth provider and implementations vary (e.g. some allow wildcards / host-based lists and some exact urls). is there a (non-scam) use case for the redirect uri not being on the host the lemmy instance is on?

Reply:
OAuth specifies that redirect_uri must be strictly compared against the registered redirect_uris for the application using simple string comparison.

don't allow wild cards because a) non-spec, b) wildly unsafe.

Reply:
I'm specifically going to refer you to the OAuth 2.0 Security Best Practices document: https://www.ietf.org/archive/id/draft-ietf-oauth-security-topics-27.html#name-insufficient-redirect-uri-v

Which recommends prohibiting wildcards in redirect_uris

Reply:
> In the end, to support having multiple clients, we completely dropped the validation at the Lemmy side and relied on the validation of the provider

This is a good point and since having a good flow for third party clients is probably pretty complicated, it's probably better to leave it up to instance admins / app developers for now and leave the redirect url unchecked.